crates/bevy_ecs/src/resource/resources.rs

```diff
@@ -454,9 +461,19 @@ pub struct ResourceRefMut<'a, T: 'static> {
 
 impl<'a, T: 'static> ResourceRefMut<'a, T> {
     /// Creates a new entity component mutable borrow
-    pub fn new(resource: &'a mut T, borrow: &'a AtomicBorrow) -> Self {
-        borrow.borrow_mut();
-        Self { resource, borrow }
+    pub fn new(resource: &'a std::cell::UnsafeCell<T>, borrow: &'a AtomicBorrow) -> Self {
```

Comment:
This function needs to be unsafe, and we need to document that the borrow is for the value.

Additionally, I'm not sure if the semantics of UnsafeCell is the right function argument, my gut feeling is that `*mut T` would be better, although I'm not sure

Reply:
> Might be good to run it through Miri

I tried doing this but none of the tests seemed to run. Just tried again after cd'ing directly into bevy_ecs and managed to get tests running, this passes under miri.
``test resource::resources::tests::thread_local_resource_mut_ref_aliasing ... ok``

> This function needs to be unsafe, and we need to document that the borrow is for the value.

Right this totally slipped my mind, one thing I can do to address both this comment and the next is change new to take a &StoredResource, it would make it a completely safe abstraction with no unsafe fn required but we'd have to make the function non-pub. I don't think this is a huge deal as those get and get_mut methods seem to be the only thing interacting with this fn.